### DISCUSSION TITLE
Color in "web" display mode

### DISCUSSION THREAD

**Discussion by Yarkane** [jimy-byerley/pymadcad#48] [Ideas]
Hey there,

We got an idea of an esthetic feature that could be great to implement (or that is already implemented but we don't know how to do).

We already know we can choose a color for each mesh.

![image](https://user-images.githubusercontent.com/45920565/191768847-57f949b7-b7f5-4535-bd03-ca80bb886a7f.png)

In our project, we implemented a "wireframe" mode, in which we remove the faces.

![image](https://user-images.githubusercontent.com/45920565/191768974-235caa2b-5ca2-4348-8b09-34e20d292da3.png)

You can see there the sphere, which is unselected, and the cylinder, which is selected.
You can also see that the color of the cylinder does not correspond to the color of the mesh, it's green whatever we select.

Is there a way to implement an option that allows us to have the same color, or, for each item, to define a color of "selection filter" ?

We select the clicked item with this code :

```
item = self.scene.displays[key]
if type(item).__name__ in ('SolidDisplay', 'WebDisplay'):
    item.vertices.selectsub(0)
```

We switch from "solid" to "web" display mode with this code :

```
self.scene.options['display_faces'] = (self.display_mode == DisplayMode.MESH)
self.scene.dequeue()
self.repaint()
```

Have a good day !

**Comment by jimy-byerley** [comment]
Hello there !
So if I clearly understood:
- you need the wireframe mode to use the same mesh color as the shaded surface mode
- alternatively you need a selection color per Mesh in addition to the existing non-selection color per Mesh

The first option in fact is already implemented, but maybe not in the way you would think. The actual wireframe-mode color is a mix between the color of lines in settings and the color of the Mesh. You can check it [here](https://github.com/jimy-byerley/pymadcad/blob/a45f53b99171887a5f23c66d4a4d4fdbc10b75b0/madcad/displays.py#L253). I choose that mix because taking the exact same color as the surface doesn't make the mesh edges that visible on top of the mesh with the same color. And I kept the same mix for the wireframe mode.
Well It shouldn't be difficult to use the mesh color for both shaded and wireframe mode. but such color could not be suited readability. For instance you can have in shaded mode a white window background, light surfaces and dark edges; then switched to wireframe mode you would have light surfaces and lines on top of a white background ...

For the second option I think that having a different color-while-selected each Mesh would be confusing for the end-user. Having the same highlight color for any selected object makes it easy to recognize when an object is selected or not, don't you think ?

**Comment by Yarkane** [comment]
Thanks for your answer ! We totally agree for the second point, my question was more about the first one.

I understand your choices of design, I'll maybe try by myself to implement my idea and see how much unreadable it is ahah.

Thanks again for your hard work !

**Reply by jimy-byerley** [reply]
Okay, let me know then
It should be fairly simple as you only need to change few mixes of this kind in `displays.SolidDisplay` and maybe  the `SolidDisplay.stack()` method

**Reply by jimy-byerley** [reply]
I've been playing with the color mix a bit, and it seems that the following gives quite readable results
```python
s = settings.display
line = (	(length(s['line_color']) + 2*length(color-s['solid_color'])) 
		/ length2(color + 0.5*s['background_color']) 
		* normalize(color + 1e-6)  )
```

**Reply by jimy-byerley** [reply]
This seems to be even better
```python
s = settings.display
line = (	(length(s['line_color']) + dot(color-s['solid_color'], s['line_color']-s['solid_color']))
		* normalize(color + 1e-6)  )
```

**Comment by Yarkane** [comment]
Hi there !

I let this idea in the back of my mind, working on the application then I tried another way to implement what I want to do, and it seems closer than what I wanted !
![image](https://user-images.githubusercontent.com/45920565/204014960-3c397a2c-2151-446b-9d73-f3224333680e.png)

To achieve this, here is a bit of my application code :

```python
def change_display_mode(self):
        if self.display_mode == DisplayMode.MESH:
            self.display_mode = DisplayMode.WEB
        else:
            self.display_mode = DisplayMode.MESH

        # Display facets or not
        for key in self._shapes3d:
            mesh : Mesh = self._shapes3d[key]
            color = mesh.options["color"]

            if self.display_mode == DisplayMode.WEB:
                new_mesh = madcad.Mesh(
                    mesh.points, 
                    mesh.faces, 
                    tracks = range(len(mesh.faces)), 
                    groups = mesh.faces,
                    )
            else:
                new_mesh = madcad.Mesh(
                    mesh.points, 
                    mesh.faces
                    )
            
            new_mesh.option(color = color)
            self.update_mesh(
                    new_mesh,
                    key,
                    f"{key}_bounds" in self.scene.displays, # Display the bounds only if that is needed
                    True
                )

        self.refresh_3d(center=False)
        self.scene.sync(self.scene.displays)

    def refresh_3d(self, center=True):
        self.scene.options['display_faces'] = (
            self.display_mode == DisplayMode.MESH)
        self.scene.dequeue()
        if center:
            _interest = self.scene.box()
            self.adjust(_interest)
            self.center(_interest.center)
        self.repaint()

```

As simple as that !

Have a good day !
